gc/src/weak/ephemeron.rs

```diff
@@ -0,0 +1,234 @@
+//! This module will implement the internal types GcBox and Ephemeron
+use crate::gc::{finalizer_safe, GcBox, GcBoxType};
+use crate::trace::Trace;
+use crate::{clear_root_bit, set_data_ptr, Finalize, GcPointer};
+use std::cell::Cell;
+use std::mem;
+use std::ptr::NonNull;
+
+/// Implementation of an Ephemeron structure
+///
+/// An Ephemeron can be either a WeakPair (Ephemeron<K,V>) or a WeakBox (Ephemeron<K,()>)
+///
+///
+/// # Tracing with Ephemerons
+///
+/// Tracing with ephemerons requires a 3 phase approach:
+///   - Phase One: Trace everything up to an ephemeron (queue found ephemerons)
+///   - Phase Two: Trace keys of queued ephemerons. If reachable,
+///
+/// [Reference]: https://docs.racket-lang.org/reference/ephemerons.html#%28tech._ephemeron%29
+pub struct Ephemeron<T: Trace + ?Sized + 'static> {
+    key: Cell<NonNull<GcBox<T>>>,
+    value: Cell<Option<NonNull<GcBox<T>>>>,
+}
```

Comment:
`Ephemeron` should have _two_ type parameters `Ephemeron<K, V>`, not one type parameter `Ephemeron<T>`. I think it should look more like

```rust
pub struct Ephemeron<K: Trace + ?Sized + 'static, V: Trace + ?Sized + 'static> {
    key: Cell<NonNull<GcBox<K>>>,
    value: V,
}
```

so that it’s up to the user whether `value` is of type `()` or `Option<…>` or `Gc<…>` or anything else.

Reply:
This makes sense, but shouldn't `Ephemeron`'s value field be an Option<V>, because we probably need to be able to set the value to None. I think is a change that needs to be added to `Ephemeron`'s Finalizer trait, so that when an ephemeron key isn't reachable and the ephemeron is finalized, we set value to None?